dalle2_pytorch/train_diffusion_prior.py

```diff
@@ -0,0 +1,160 @@
+                                     timesteps = 100, cond_drop_prob = 0.2, loss_type = loss_type,
+                                     condition_on_text_encodings = condition_on_text_encodings).to(device)
+    # Gtext_reader image and text embeddings from the servers
+    print("==============Downloading embeddings - image and text====================")
+    image_reader = EmbeddingReader(embeddings_folder=image_embed_url, file_format="npy")
+    text_reader = EmbeddingReader(embeddings_folder=text_embed_url, file_format="npy")
+
+    ### Training code ###
+    optimizer = torch.optim.SGD(diffusion_prior.parameters(), lr = learning_rate)
+    epochs = num_epochs
+    min_valid_loss = np.inf
+    for e in range(epochs):
+        train_loss = 0.0 
+        print("Training loop - epoch number ",e)
+        train_set_size = int(train_percent*image_reader.count)
+        for emb_images,emb_text in zip(image_reader(batch_size=batch_size, start=0, end=train_set_size),
+                text_reader(batch_size=batch_size, start=0, end=train_set_size)):
+            emb_images = list(emb_images)
+            emb_text = list(emb_text)
+            print(emb_images[0].shape,emb_text[0].shape)
+            if torch.cuda.is_available():
+                emb_images[0] = torch.tensor(emb_images[0]).to(device)
+                emb_text[0] = torch.tensor(emb_text[0]).to(device)
+            optimizer.zero_grad()
+            loss = diffusion_prior(text_embed = emb_text[0],image_embed = emb_images[0])
+            loss.backward()
+            # Log to wandb
+            wandb.log({"Training (mse)": loss})
+            optimizer.step()
+            print("Training mse = ",loss.item())
+            train_loss+=loss.item()
+
+        print("Validation loop - epoch number ",e)
+        with torch.no_grad():
+            valid_loss = 0.0
+            val_set_size = int(val_percent*image_reader.count)
+            start = train_set_size
+            end=start+val_set_size
+            for emb_images,emb_text in zip(image_reader(batch_size=batch_size, start=start, end=end),
+                    text_reader(batch_size=batch_size, start=start, end=end)):
+                emb_images = list(emb_images)
+                emb_text = list(emb_text)
+                if torch.cuda.is_available():
+                    emb_images[0] = torch.tensor(emb_images[0]).to(device)
+                    emb_text[0] = torch.tensor(emb_text[0]).to(device)
+                diffusion_prior.eval()
+                loss = diffusion_prior(text_embed = emb_text[0],image_embed = emb_images[0])
+
+                # Log to wandb
+                wandb.log({"Validation Loss(mse) ": loss})
+                valid_loss+=loss.item()
+
+        # Saving State Dict
+        torch.save(diffusion_prior.state_dict(), 'saved_model.pth')
+
+    test_set_size = int(test_percent*train_set_size)
+    with torch.no_grad():
```

Comment:
I would recommend to put that code in a function that can be reused for both validation and test (with different start and end settings)